### ISSUE TITLE
pgbouncer didn't even try auth_query to verify user password

### ISSUE THREAD

**Issue by bagafoot** [pgbouncer/pgbouncer#484]
Hi all, I'm stuck and don't know where I'm doing wrong.
I'm try to configure pgbouncer to auth_type = hba and auth_user with auth_query but pgbouncer didn't send request to database. Pgbouncer and Postgres on the same server. What I'm missing

```
pgbouncer --version
PgBouncer 1.13.0
libevent 2.0.21-stable
adns: libc-2.17
tls: OpenSSL 1.0.2k-fips  26 Jan 2017
PostgreSQL 12.2 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39), 64-bit
```
pgbouncer.ini

```
[databases]
test = dbname=test host=127.0.0.1 port=5432
test = auth_user=pgbouncer
[users]
test_app = pool_mode=session max_user_connections=3
[pgbouncer]
logfile = /var/log/pgbouncer/pgbouncer.log
pidfile = /var/run/pgbouncer/pgbouncer.pid
listen_addr = 127.0.0.1
listen_port = 6432
auth_type = hba
auth_file = /etc/pgbouncer/userlist.txt
auth_hba_file = /etc/pgbouncer/pg_hba.conf
auth_query = SELECT p_user, p_password FROM public.lookup($1)
admin_users = pgbouncer
stats_users = stats
server_reset_query = DISCARD ALL
server_reset_query_always = 1
max_client_conn = 100
default_pool_size = 20
log_connections = 1
log_disconnections = 1
log_pooler_errors = 1
log_stats = 1
verbose = 1
tcp_keepalive = 1
job_name = pgbouncer

```
userlist file
`"pgbouncer" "md5be5544d3807b54dd0637f2439ecb03b9"`

pgbouncer pg_hba.conf file
```
host    all     pgbouncer       127.0.0.1/32    trust
host    test    test_app        127.0.0.1/32    md5
```

postgres pg_hba.conf file
```
 # TYPE  DATABASE        USER            ADDRESS                 METHOD
 
 # "local" is for Unix domain socket connections only
 local   all             all                                    trust
 # IPv4 local connections:
 host    all             all             127.0.0.1/32            md5
```

lookup function 
```
test=# \df
                                 List of functions
 Schema |  Name  | Result data type |          Argument data types           | Type
--------+--------+------------------+----------------------------------------+------
 public | lookup | record           | INOUT p_user name, OUT p_password text | func
```

**Comment by petere** [comment]
You also need to tell us how you attempted to connect (which command and which options), and the log entries in the postgres and pgbouncer logs from around that time.

**Comment by bagafoot** [comment]
> You also need to tell us how you attempted to connect (which command and which options), and the log entries in the postgres and pgbouncer logs from around that time.

psql commands
![testpsql](https://user-images.githubusercontent.com/30403489/82058097-f53afd00-96cc-11ea-9cf2-a526fd675584.PNG)

and tail -f  db log file 

![dblog](https://user-images.githubusercontent.com/30403489/82058205-18fe4300-96cd-11ea-8af1-21eabac0827b.PNG)

**Comment by petere** [comment]
This setup

```
[databases]
test = dbname=test host=127.0.0.1 port=5432
test = auth_user=pgbouncer
```

is not correct. You have two entries for the `test` database. You need to combine those into one.

**Comment by bagafoot** [comment]
> This setup
> 
> ```
> [databases]
> test = dbname=test host=127.0.0.1 port=5432
> test = auth_user=pgbouncer
> ```
> 
> is not correct. You have two entries for the `test` database. You need to combine those into one.

I've change configuration and remove  `test = auth_user=pgbouncer` line and restart/reload pgbouncer  but still the same error.


![config](https://user-images.githubusercontent.com/30403489/82125653-03157e80-97b0-11ea-8bdc-a8ef5af4de91.PNG)

![reload](https://user-images.githubusercontent.com/30403489/82125659-0c9ee680-97b0-11ea-8251-733ad4843d8a.PNG)

![errror](https://user-images.githubusercontent.com/30403489/82125663-13c5f480-97b0-11ea-80b8-1783005a9e0b.PNG)

**Comment by eulerto** [comment]
I tried your setup and any user into _users_ section cannot login. Looking at the postgres logs, pgbouncer does not try to connect and execute _auth_query_ when user is from _users_ section. Other users (that is not in users section) work fine. It seems a bug to me.

**Comment by bagafoot** [comment]
> I tried your setup and any user into _users_ section cannot login. Looking at the postgres logs, pgbouncer does not try to connect and execute _auth_query_ when user is from _users_ section. Other users (that is not in users section) work fine. It seems a bug to me.

You are right, I tried too after your comment and It works fine after remove users in `[users]` section.

**Comment by sribe-drw** [comment]
Yes, I just went through this today, and have some more details. In my specific case I set `client_tls_sslmode = prefer` in the `[pgbouncer]` section. What happens then is that any user that has settings specified in the `[users]` section ignores that, uses the default of no SSL, does not even attempt to fetch user info via the `auth_query` and reports that password auth failed. 

I find that the statement in the docs "Only a few settings are available here." followed by just the two `pool_mode` and `max_user_connections` is misleading. In fact, adding `client_tls_sslmode=prefer` to the user's entry fixes the problem. (That is all that is necessary; I did not repeat the settings for cert/key locations and it worked.)

**Comment by sribe-drw** [comment]
Addendum to above: reviewing logs from after out config changes, I see:

```
ERROR skipping user ... because of unknown parameter in settings: client_tls_sslmode
```

So in fact it "works" by ignoring the setting where we try to limit the pool size for that user, so there seems to be NO WAY to have TLS and per-user settings.

**Comment by jpuris** [comment]
Spent better half of the day on this and can confirm that with following config configuration will not let `postgre` user to connect to DB, unless it is listed in `userlist.txt`
```ini
[databases]

pgbench = host=127.0.0.1 port=5431 auth_user=pgbouncer pool_size=4
postgres = host=127.0.0.1 port=5431 auth_user=pgbouncer pool_size=0

[users]
postgres = max_user_connections=2

[pgbouncer]
listen_addr = *
listen_port = 5432

logfile = /var/log/pgbouncer/pgbouncer.log
pidfile = /var/run/pgbouncer/pgbouncer.pid

unix_socket_dir = /var/run/pgbouncer

auth_type = md5
auth_query = SELECT p_user, p_password FROM pgbouncer.auth($1)
auth_file = /etc/pgbouncer/userlist.txt

pool_mode = session
server_reset_query = DISCARD ALL

stats_users = pgbouncer_stats
admin_users = pgbouncer_admin
```
When `[users]` section is removed and pgbouncer is _**restarted**_ (reload will not work for this), postgres user is able to log in via pgbouncer.

This pretty much hard blocks us to use `[users]` all together and there is no way for us to add connection limits to particular users.

**Comment by petere** [comment]
It has become apparent, also through other reports, that the `[users]` section support is pretty buggy. It's good to have this reproducible report. The whole thing probably needs a deeper analysis and rewrite.

**Comment by Justin-Kwan** [comment]
@eulerto @petere Here is a PR to fix this issue: https://github.com/pgbouncer/pgbouncer/pull/706

**Comment by veshant** [comment]
> @eulerto @petere Here is a PR to fix this issue: #706

@petere can this PR be merged in? I have the same issue

**Comment by JelteF** [comment]
@veshant The #706 PR is not in a state where it can be merged.
